File: calclub/lib/domain/entities/health_data.dart
import 'package:equatable/equatable.dart';

/// Clean domain entity for daily steps data
class DailyStepsData extends Equatable {
  final int totalSteps;
  final DateTime date;
  final DateTime lastUpdated;

  const DailyStepsData({
    required this.totalSteps,
    required this.date,
    required this.lastUpdated,
  });

  @override
  List<Object?> get props => [totalSteps, date, lastUpdated];
}

/// Clean domain entity for workout/exercise data
class WorkoutData extends Equatable {
  final String id;
  final WorkoutType workoutType;
  final double activeCalories;
  final Duration duration;
  final DateTime startTime;
  final DateTime endTime;
  final int? steps;
  final double? distance; // in meters
  final double? averageHeartRate;
  final double? peakHeartRate;
  final double? averagePace; // in minutes per km

  const WorkoutData({
    required this.id,
    required this.workoutType,
    required this.activeCalories,
    required this.duration,
    required this.startTime,
    required this.endTime,
    this.steps,
    this.distance,
    this.averageHeartRate,
    this.peakHeartRate,
    this.averagePace,
  });

  @override
  List<Object?> get props => [
        id,
        workoutType,
        activeCalories,
        duration,
        startTime,
        endTime,
        steps,
        distance,
        averageHeartRate,
        peakHeartRate,
        averagePace,
      ];
}

/// Enum for different workout types
enum WorkoutType {
  running,
  walking,
  cycling,
  swimming,
  strengthTraining,
  yoga,
  hiking,
  other;

  String get displayName {
    switch (this) {
      case WorkoutType.running:
        return 'Running';
      case WorkoutType.walking:
        return 'Walking';
      case WorkoutType.cycling:
        return 'Cycling';
      case WorkoutType.swimming:
        return 'Swimming';
      case WorkoutType.strengthTraining:
        return 'Strength Training';
      case WorkoutType.yoga:
        return 'Yoga';
      case WorkoutType.hiking:
        return 'Hiking';
      case WorkoutType.other:
        return 'Exercise';
    }
  }

  String get icon {
    switch (this) {
      case WorkoutType.running:
        return '🏃';
      case WorkoutType.walking:
        return '🚶';
      case WorkoutType.cycling:
        return '🚴';
      case WorkoutType.swimming:
        return '🏊';
      case WorkoutType.strengthTraining:
        return '💪';
      case WorkoutType.yoga:
        return '🧘';
      case WorkoutType.hiking:
        return '🥾';
      case WorkoutType.other:
        return '⚡';
    }
  }
}

File: backend/models/HealthData.js
const mongoose = require('mongoose');

// Schema for daily steps data
const DailyStepsSchema = new mongoose.Schema({
  userId: {
    type: String,
    required: true,
    index: true
  },
  totalSteps: {
    type: Number,
    required: true
  },
  date: {
    type: Date,
    required: true
  },
  lastUpdated: {
    type: Date,
    required: true
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
});

// Compound index for efficient queries
DailyStepsSchema.index({ userId: 1, date: 1 }, { unique: true });

// Schema for workout data
const WorkoutSchema = new mongoose.Schema({
  userId: {
    type: String,
    required: true,
    index: true
  },
  workoutId: {
    type: String,
    required: true,
    unique: true
  },
  workoutType: {
    type: String,
    required: true
  },
  activeCalories: {
    type: Number,
    required: true
  },
  durationMinutes: {
    type: Number,
    required: true
  },
  startTime: {
    type: Date,
    required: true
  },
  endTime: {
    type: Date,
    required: true
  },
  steps: {
    type: Number,
    default: null
  },
  distance: {
    type: Number,
    default: null
  },
  averageHeartRate: {
    type: Number,
    default: null
  },
  peakHeartRate: {
    type: Number,
    default: null
  },
  averagePace: {
    type: Number,
    default: null
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
});

// Index for efficient queries
WorkoutSchema.index({ userId: 1, startTime: -1 });

// Create models
const DailySteps = mongoose.model('DailySteps', DailyStepsSchema);
const Workout = mongoose.model('Workout', WorkoutSchema);

module.exports = {
  DailySteps,
  Workout
};